table/block.cc

```diff
@@ -170,6 +170,14 @@ void DataBlockIter::Seek(const Slice& target) {
     return;
   }
   uint32_t index = 0;
+
+  if (data_block_hash_index_) {
+    // suffix seek will set the current_ and restart_index_,
+    // no need to pass back `index`, or linear search.
+    HashSeek(target);
```

Comment:
I don't see how this hash seek is avoided for range lookups.

Reply:
Thanks. This is fixed by adding a bool parameter `is_point_look_up` to `Block::NewIterator()`. More details can be found in the commit here: https://github.com/facebook/rocksdb/pull/4174/commits/9c0646cd1dc9255fd101d5ed25e5268beac650ac.

Reply:
The data format of the DataBlockHashIndex can be found in here: https://github.com/facebook/rocksdb/blob/master/table/data_block_hash_index.h#L11-L74. Do we also want to include this description in `block.cc`? Or we just write a comment saying more details please refer to `table/data_block_hash_index.h`?

Reply:
If it is already documented, it is fine then.

Reply:
@fgwu this interface is confusing to me: an **iterator** is "point look up". What's the reason that you want to fit the logic into DataBlockIter? Is it because you want to make this for-loop work? https://github.com/facebook/rocksdb/blob/v5.14.2/table/block_based_table_reader.cc#L2116-L2129

To me, it is debatable which one is easier to change to make work, DataBlockIter, or those lines.

Reply:
@siying A fallback logic implementation using DataBlockIter is added to this PR, where `BlockBasedTable:: NewDataBlockIterator()` takes one argument `is_data_block_point_lookup ` to constructor a `biter` instance.

Logically, a `Block::Get()` is a straightforward way to support `BlockBasedTable::Get`. However, before `Block::Get()` is called, there are many preparation procedures, such as checking block cache, reading file, update statistics, cleanup on any failure, etc. Such code path already exists in [`BlockBasedTable:: NewDataBlockIterator()`](https://github.com/facebook/rocksdb/blob/v5.14.2/table/block_based_table_reader.cc#L1481-L1483).  

We can possibly extract the preparation logic and put them into a dedicated preparation function for `Block::Get()`. However, given that our HashSeek fall back often to binary on collision, `BlockBasedTable:: NewDataBlockIterator()` is still called when we fall back to use binary seek. Here the problem is, the same preparation is performed twice: one is done when prepare for Block::Get(), the other is done in `BlockBasedTable:: NewDataBlockIterator()`. 

For example `MaybeLoadDataBlockToCache()` will be called twice, causing extra CPU consumption and doubles the statistics for operations. If we add some if-else in `BlockBasedTable:: NewDataBlockIterator()` to avoid such repetition, we are essentially  refactoring `BlockBasedTable:: NewDataBlockIterator()`.

Still, we can extract the preparation routines from `BlockBasedTable:: NewDataBlockIterator()` to a separate function. But refactoring `BlockBasedTable:: NewDataBlockIterator()` requires some non-trivial code change. 

Yet another possibility is to reuse `BlockBasedTable:: NewDataBlockIterator()` with minimal change. It only creates the `biter` differently. 

Therefore, `BlockBasedTable:: NewDataBlockIterator()`  is only called once to do the preparation.  `biter` firstly tries `HashSeek`. If it fails and falls back, we change the internal state of the `biter` and reuse the `biter` to perform binary seek.